src/sparse_blas/backends/rocsparse/rocsparse_scope_handle.cpp

```diff
@@ -0,0 +1,135 @@
+        // No context is installed and the suggested context is primary
+        // This is the most common case. We can activate the context in the
+        // thread and leave it there until all the PI context referring to the
+        // same underlying rocsparse primary context are destroyed. This emulates
+        // the behaviour of the rocsparse runtime api, and avoids costly context
+        // switches. No action is required on this side of the if.
+        needToRecover_ = !(original_ == nullptr);
+    }
+}
+
+RocsparseScopedContextHandler::~RocsparseScopedContextHandler() noexcept(false) {
+    if (needToRecover_) {
+        HIP_ERROR_FUNC(hipCtxSetCurrent, original_);
+    }
+    delete placedContext_;
+}
+
+void ContextCallback(void* userData) {
+    auto* atomic_ptr = static_cast<std::atomic<rocsparse_handle>*>(userData);
+    if (!atomic_ptr) {
+        return;
+    }
+    auto handle = atomic_ptr->exchange(nullptr);
+    if (handle != nullptr) {
+        ROCSPARSE_ERR_FUNC(rocsparse_destroy_handle, handle);
+    }
+    delete atomic_ptr;
+}
+
+std::pair<rocsparse_handle, hipStream_t> RocsparseScopedContextHandler::get_handle_and_stream(
+    const sycl::queue& queue) {
+    auto hipDevice = ih.get_native_device<sycl::backend::ext_oneapi_hip>();
+    hipCtx_t desired;
+    HIP_ERROR_FUNC(hipDevicePrimaryCtxRetain, &desired, hipDevice);
+#ifdef ONEAPI_ONEMKL_PI_INTERFACE_REMOVED
+    auto piPlacedContext_ = reinterpret_cast<ur_context_handle_t>(desired);
+#else
+    auto piPlacedContext_ = reinterpret_cast<pi_context>(desired);
+#endif
+    hipStream_t streamId = sycl::get_native<sycl::backend::ext_oneapi_hip>(queue);
+    auto it = handle_helper.rocsparse_handle_container_mapper_.find(piPlacedContext_);
+    if (it != handle_helper.rocsparse_handle_container_mapper_.end()) {
+        if (it->second == nullptr) {
+            handle_helper.rocsparse_handle_container_mapper_.erase(it);
+        }
+        else {
+            auto handle = it->second->load();
+            if (handle != nullptr) {
+                hipStream_t currentStreamId;
+                ROCSPARSE_ERR_FUNC(rocsparse_get_stream, handle, &currentStreamId);
+                if (currentStreamId != streamId) {
+                    ROCSPARSE_ERR_FUNC(rocsparse_set_stream, handle, streamId);
+                }
+                return { handle, streamId };
+            }
+            else {
+                handle_helper.rocsparse_handle_container_mapper_.erase(it);
+            }
+        }
+    }
+
+    rocsparse_handle handle;
+    ROCSPARSE_ERR_FUNC(rocsparse_create_handle, &handle);
+    ROCSPARSE_ERR_FUNC(rocsparse_set_stream, handle, streamId);
+
+    auto atomic_ptr = new std::atomic<rocsparse_handle>(handle);
+    handle_helper.rocsparse_handle_container_mapper_.insert(
+        std::make_pair(piPlacedContext_, atomic_ptr));
+
+    sycl::detail::pi::contextSetExtendedDeleter(*placedContext_, ContextCallback, atomic_ptr);
```

Comment:
Is it unchanged between the removed PI versus new UR? Isn't there a `sycl::detail::ur::contextSetExtendedDeleter` somewhere?